In [24]:
#import torch
#from torchvision import datasets
#from torch.utils.data import DataLoader
#from torchvision import transforms
import numpy as np
from utilsCPD import *
import matplotlib.pyplot as plt
from SWCPD import BaseDetector as SWDCP
import os
from pathlib import Path

In [25]:
file_name = 'Occupancy.csv'

file_path = os.path.normpath(os.path.join(os.getcwd(),"datasets/Occupancy",file_name))

Occupancy = pd.read_csv(file_path).iloc[:,1:]

Targets = pd.read_csv(os.path.normpath(os.path.join(os.getcwd(),"datasets/Occupancy/OccupancyTargets.csv"))).iloc[:,1:]

GroundTruth = list(Targets.to_numpy().flatten())

Occupancy.head()

,Temperature,Humidity,Light,CO2
0,23.18,27.2720,426.0,721.25
1,23.15,27.2675,429.5,714.00
2,23.15,27.2450,426.0,713.50
3,23.15,27.2000,426.0,708.25
4,23.10,27.2000,426.0,704.50


In [39]:
Occupancy = (Occupancy-Occupancy.mean())/Occupancy.std()

### SWCPD

In [40]:
detector = SWDCP(data=Occupancy.to_numpy().astype(np.float32),window_length=500,max_history=500,significance=0.05)
print(detector.device)
detector.process_dataloader(p=2,n_theta=1000,seed=2025)
detector.evaluate(GroundTruth,30)

cuda


  0%|          | 0/7644 [00:00<?, ?it/s]

  5%|▍         | 365/7644 [00:03<01:23, 87.50it/s]

Change detected at: 851 
Initiate new segment


 14%|█▎        | 1050/7644 [00:13<03:08, 35.00it/s]

Change detected at: 1540 
Initiate new segment


 24%|██▎       | 1814/7644 [00:20<00:59, 98.79it/s] 

Change detected at: 2303 
Initiate new segment


 33%|███▎      | 2504/7644 [00:27<00:50, 102.40it/s]

Change detected at: 2988 
Initiate new segment


 44%|████▍     | 3349/7644 [00:36<00:48, 88.96it/s] 

Change detected at: 3831 
Initiate new segment


 65%|██████▌   | 4972/7644 [00:55<00:28, 92.25it/s] 

Change detected at: 5456 
Initiate new segment


 81%|████████  | 6169/7644 [01:11<00:18, 78.05it/s] 

Change detected at: 6656 
Initiate new segment


 91%|█████████▏| 6986/7644 [01:19<00:06, 108.47it/s]

Change detected at: 7462 
Initiate new segment


100%|█████████▉| 7615/7644 [01:26<00:00, 100.87it/s]

Change detected at: 8098 
Initiate new segment


100%|██████████| 7644/7644 [01:27<00:00, 87.56it/s] 

F1 score: 0.34782608695652173
Covering: 0.7019102351675222
0.35384615384615387


(0.34782608695652173,
 0.7019102351675222,
 0.35384615384615387,
 6,
 ({6656: 1,
   8098: 3,
   1540: 85,
   7462: 247,
   2988: 93,
   5456: 273,
   851: 20,
   3831: 88,
   2303: 32},
  93.55555555555556))

In [42]:
GroundTruth

[15, 831, 1455, 2271, 2895, 3743, 4271, 5183, 5759, 6655, 7215, 8095]

In [83]:
detector = SWDCP(data=Occupancy.to_numpy().astype(np.float32),window_length=250,max_history=250,significance=0.2,use_cuda=False)
print(detector.device)
detector.process_dataloader(p=2,n_theta=1000,seed=2025)
detector.evaluate(GroundTruth,30)

cpu


  8%|▊         | 616/7894 [00:02<00:26, 269.70it/s]

Change detected at: 828 
Initiate new segment


 16%|█▌        | 1262/7894 [00:04<00:22, 290.79it/s]

Change detected at: 1469 
Initiate new segment


 23%|██▎       | 1836/7894 [00:06<00:20, 293.45it/s]

Change detected at: 2028 
Initiate new segment


 26%|██▌       | 2068/7894 [00:07<00:20, 282.06it/s]

Change detected at: 2271 
Initiate new segment


 32%|███▏      | 2548/7894 [00:09<00:20, 262.32it/s]

Change detected at: 2759 
Initiate new segment


 34%|███▍      | 2708/7894 [00:09<00:21, 246.61it/s]

Change detected at: 2915 
Initiate new segment


 41%|████      | 3240/7894 [00:11<00:15, 294.38it/s]

Change detected at: 3458 
Initiate new segment


 45%|████▌     | 3576/7894 [00:12<00:14, 295.92it/s]

Change detected at: 3790 
Initiate new segment


 51%|█████     | 4041/7894 [00:14<00:12, 311.27it/s]

Change detected at: 4251 
Initiate new segment


 64%|██████▍   | 5060/7894 [00:17<00:09, 296.29it/s]

Change detected at: 5265 
Initiate new segment


 67%|██████▋   | 5250/7894 [00:18<00:08, 312.43it/s]

Change detected at: 5436 
Initiate new segment


 76%|███████▌  | 6000/7894 [00:20<00:06, 295.85it/s]

Change detected at: 6213 
Initiate new segment
Change detected at: 6240 
Initiate new segment


 80%|████████  | 6352/7894 [00:21<00:04, 321.96it/s]

Change detected at: 6554 
Initiate new segment
Change detected at: 6602 
Initiate new segment


 82%|████████▏ | 6449/7894 [00:21<00:04, 311.64it/s]

Change detected at: 6655 
Initiate new segment


 87%|████████▋ | 6868/7894 [00:23<00:03, 333.36it/s]

Change detected at: 7096 
Initiate new segment


 89%|████████▉ | 7060/7894 [00:23<00:02, 295.97it/s]

Change detected at: 7272 
Initiate new segment


 99%|█████████▉| 7851/7894 [00:27<00:00, 253.76it/s]

Change detected at: 8061 
Initiate new segment


100%|██████████| 7894/7894 [00:27<00:00, 289.73it/s]

F1 score: 0.4242424242424242
Covering: 0.7429371956779628
0.4442307692307692


(0.4242424242424242,
 0.7429371956779628,
 0.4442307692307692,
 13,
 ({3458: 285,
   5265: 82,
   6554: 101,
   4251: 20,
   7096: 119,
   5436: 253,
   828: 3,
   1469: 14,
   6213: 442,
   2759: 136,
   6602: 53,
   3790: 47,
   2271: 0,
   6240: 415,
   2915: 20,
   7272: 57,
   2028: 243,
   8061: 34,
   6655: 0},
  122.3157894736842))

|$K_{max}$|$\alpha=0.01$|$\alpha=0.05$|$\alpha=0.1$|$\alpha=0.2$|
|---|---|---|---|---|
|$50$|$2$|$5$|$8$|$13$|
|$100$|$5$|$7$|$11$|$15$|
|$200$|$4$|$7$|$7$|$14$|
|$250$|$5$|$6$|$6$|$13$|

In [37]:
FPs = []

for w in [100,200,400,500]:
    for frac in [0.5,1.0]:
        detector = SWDCP(data=Occupancy.to_numpy().astype(np.float32),window_length=w,max_history=int(w*frac),significance=0.01)
        print(detector.device)
        detector.process_dataloader(p=2,n_theta=1000,seed=2025)
        _,_,_,fp,_ =detector.evaluate(GroundTruth,30)
        FPs.append(fp)


cuda


  9%|▉         | 754/8044 [00:06<00:58, 125.54it/s]

Change detected at: 828 
Initiate new segment


 12%|█▏        | 981/8044 [00:07<00:56, 124.79it/s]

Change detected at: 1068 
Initiate new segment


 14%|█▎        | 1100/8044 [00:08<00:59, 116.87it/s]

Change detected at: 1180 
Initiate new segment


 17%|█▋        | 1342/8044 [00:12<01:06, 101.25it/s]

Change detected at: 1423 
Initiate new segment


 23%|██▎       | 1810/8044 [00:16<00:46, 135.02it/s]

Change detected at: 1884 
Initiate new segment


 26%|██▌       | 2062/8044 [00:18<00:44, 132.94it/s]

Change detected at: 2140 
Initiate new segment
Change detected at: 2144 
Initiate new segment


 26%|██▋       | 2118/8044 [00:18<00:46, 126.95it/s]

Change detected at: 2191 
Initiate new segment
Change detected at: 2194 
Initiate new segment
Change detected at: 2203 
Initiate new segment


 27%|██▋       | 2201/8044 [00:19<00:44, 132.48it/s]

Change detected at: 2270 
Initiate new segment


 31%|███       | 2498/8044 [00:21<00:42, 129.26it/s]

Change detected at: 2585 
Initiate new segment


 35%|███▍      | 2808/8044 [00:23<00:39, 133.04it/s]

Change detected at: 2892 
Initiate new segment


 45%|████▍     | 3581/8044 [00:29<00:28, 156.09it/s]

Change detected at: 3649 
Initiate new segment


 46%|████▌     | 3674/8044 [00:29<00:30, 141.69it/s]

Change detected at: 3760 
Initiate new segment


 49%|████▉     | 3927/8044 [00:31<00:31, 132.69it/s]

Change detected at: 4002 
Initiate new segment


 54%|█████▍    | 4363/8044 [00:35<00:27, 132.67it/s]

Change detected at: 4437 
Initiate new segment


 56%|█████▋    | 4532/8044 [00:36<00:26, 134.30it/s]

Change detected at: 4616 
Initiate new segment


 62%|██████▏   | 4948/8044 [00:39<00:22, 138.66it/s]

Change detected at: 5034 
Initiate new segment
Change detected at: 5043 
Initiate new segment


 63%|██████▎   | 5032/8044 [00:40<00:23, 128.89it/s]

Change detected at: 5113 
Initiate new segment


 63%|██████▎   | 5103/8044 [00:40<00:23, 124.18it/s]

Change detected at: 5187 
Initiate new segment


 64%|██████▍   | 5188/8044 [00:41<00:22, 127.97it/s]

Change detected at: 5264 
Initiate new segment


 66%|██████▋   | 5330/8044 [00:42<00:20, 131.97it/s]

Change detected at: 5412 
Initiate new segment


 75%|███████▌  | 6051/8044 [00:49<00:14, 139.05it/s]

Change detected at: 6131 
Initiate new segment


 77%|███████▋  | 6196/8044 [00:50<00:14, 130.05it/s]

Change detected at: 6277 
Initiate new segment


 79%|███████▉  | 6350/8044 [00:51<00:12, 130.71it/s]

Change detected at: 6426 
Initiate new segment
Change detected at: 6428 
Initiate new segment
Change detected at: 6432 
Initiate new segment
Change detected at: 6436 
Initiate new segment


 81%|████████  | 6479/8044 [00:52<00:11, 135.95it/s]

Change detected at: 6551 
Initiate new segment
Change detected at: 6559 
Initiate new segment
Change detected at: 6565 
Initiate new segment


 82%|████████▏ | 6582/8044 [00:53<00:10, 135.27it/s]

Change detected at: 6655 
Initiate new segment


 85%|████████▍ | 6830/8044 [00:55<00:09, 131.69it/s]

Change detected at: 6906 
Initiate new segment


 89%|████████▊ | 7134/8044 [00:57<00:06, 134.23it/s]

Change detected at: 7212 
Initiate new segment


 98%|█████████▊| 7892/8044 [01:03<00:01, 122.48it/s]

Change detected at: 7977 
Initiate new segment


100%|█████████▉| 8011/8044 [01:04<00:00, 127.80it/s]

Change detected at: 8087 
Initiate new segment


100%|██████████| 8044/8044 [01:04<00:00, 123.94it/s]


F1 score: 0.34615384615384615
Covering: 0.5851391582918367
0.46153846153846156
cuda


  9%|▉         | 744/8044 [00:05<00:53, 135.48it/s]

Change detected at: 828 
Initiate new segment


 12%|█▏        | 997/8044 [00:07<00:54, 129.52it/s]

Change detected at: 1075 
Initiate new segment


 13%|█▎        | 1080/8044 [00:07<00:52, 133.69it/s]

Change detected at: 1161 
Initiate new segment


 17%|█▋        | 1342/8044 [00:09<00:48, 139.18it/s]

Change detected at: 1428 
Initiate new segment


 23%|██▎       | 1829/8044 [00:15<01:42, 60.79it/s] 

Change detected at: 1905 
Initiate new segment


 26%|██▌       | 2061/8044 [00:16<00:47, 124.86it/s]

Change detected at: 2141 
Initiate new segment
Change detected at: 2144 
Initiate new segment


 26%|██▌       | 2101/8044 [00:17<00:49, 120.64it/s]

Change detected at: 2184 
Initiate new segment


 27%|██▋       | 2184/8044 [00:17<00:43, 135.48it/s]

Change detected at: 2270 
Initiate new segment


 31%|███       | 2510/8044 [00:20<00:43, 126.08it/s]

Change detected at: 2585 
Initiate new segment


 35%|███▍      | 2804/8044 [00:22<00:44, 118.57it/s]

Change detected at: 2891 
Initiate new segment


 43%|████▎     | 3441/8044 [00:27<00:34, 134.93it/s]

Change detected at: 3524 
Initiate new segment


 45%|████▍     | 3586/8044 [00:28<00:31, 139.54it/s]

Change detected at: 3656 
Initiate new segment


 46%|████▌     | 3690/8044 [00:29<00:31, 139.89it/s]

Change detected at: 3760 
Initiate new segment


 49%|████▉     | 3928/8044 [00:30<00:29, 140.75it/s]

Change detected at: 4010 
Initiate new segment


 54%|█████▍    | 4367/8044 [00:34<00:25, 144.83it/s]

Change detected at: 4438 
Initiate new segment


 57%|█████▋    | 4547/8044 [00:35<00:25, 138.45it/s]

Change detected at: 4621 
Initiate new segment
Change detected at: 4627 
Initiate new segment


 62%|██████▏   | 4962/8044 [00:38<00:21, 143.68it/s]

Change detected at: 5043 
Initiate new segment


 63%|██████▎   | 5031/8044 [00:39<00:26, 114.13it/s]

Change detected at: 5114 
Initiate new segment


 63%|██████▎   | 5099/8044 [00:39<00:31, 93.44it/s] 

Change detected at: 5187 
Initiate new segment


 65%|██████▌   | 5235/8044 [00:41<00:31, 88.86it/s]

Change detected at: 5318 
Initiate new segment


 66%|██████▌   | 5321/8044 [00:42<00:31, 87.51it/s]

Change detected at: 5413 
Initiate new segment


 75%|███████▌  | 6047/8044 [00:50<00:20, 98.94it/s] 

Change detected at: 6130 
Initiate new segment


 77%|███████▋  | 6211/8044 [00:51<00:11, 165.94it/s]

Change detected at: 6288 
Initiate new segment


 79%|███████▉  | 6353/8044 [00:52<00:10, 167.64it/s]

Change detected at: 6426 
Initiate new segment
Change detected at: 6429 
Initiate new segment
Change detected at: 6432 
Initiate new segment
Change detected at: 6436 
Initiate new segment


 81%|████████  | 6491/8044 [00:53<00:10, 154.19it/s]

Change detected at: 6577 
Initiate new segment
Change detected at: 6582 
Initiate new segment
Change detected at: 6605 
Initiate new segment


 82%|████████▏ | 6570/8044 [00:53<00:09, 150.97it/s]

Change detected at: 6655 
Initiate new segment


 85%|████████▍ | 6828/8044 [00:55<00:07, 160.86it/s]

Change detected at: 6911 
Initiate new segment


 89%|████████▊ | 7136/8044 [00:58<00:07, 126.09it/s]

Change detected at: 7215 
Initiate new segment


 98%|█████████▊| 7894/8044 [01:04<00:01, 127.26it/s]

Change detected at: 7980 
Initiate new segment


100%|█████████▉| 8012/8044 [01:05<00:00, 119.62it/s]

Change detected at: 8087 
Initiate new segment


100%|██████████| 8044/8044 [01:05<00:00, 123.25it/s]


F1 score: 0.392156862745098
Covering: 0.5765647571268293
0.5161943319838056
cuda


  8%|▊         | 646/7944 [00:04<00:55, 131.52it/s]

Change detected at: 828 
Initiate new segment


 16%|█▌        | 1283/7944 [00:09<00:51, 130.48it/s]

Change detected at: 1464 
Initiate new segment


 26%|██▌       | 2065/7944 [00:18<05:33, 17.60it/s] 

Change detected at: 2248 
Initiate new segment
Change detected at: 2270 
Initiate new segment


 30%|███       | 2413/7944 [00:21<00:46, 118.45it/s]

Change detected at: 2589 
Initiate new segment


 34%|███▍      | 2732/7944 [00:23<00:41, 125.29it/s]

Change detected at: 2910 
Initiate new segment


 45%|████▌     | 3587/7944 [00:30<00:33, 130.03it/s]

Change detected at: 3767 
Initiate new segment


 49%|████▉     | 3892/7944 [00:32<00:30, 134.19it/s]

Change detected at: 4074 
Initiate new segment


 63%|██████▎   | 5017/7944 [00:40<00:22, 131.56it/s]

Change detected at: 5191 
Initiate new segment


 66%|██████▋   | 5267/7944 [00:42<00:23, 115.09it/s]

Change detected at: 5453 
Initiate new segment


 79%|███████▉  | 6278/7944 [00:54<00:16, 102.45it/s]

Change detected at: 6467 
Initiate new segment


 81%|████████  | 6412/7944 [00:55<00:15, 101.97it/s]

Change detected at: 6600 
Initiate new segment


 87%|████████▋ | 6910/7944 [00:59<00:09, 114.41it/s]

Change detected at: 7089 
Initiate new segment


 89%|████████▉ | 7095/7944 [01:01<00:06, 125.47it/s]

Change detected at: 7278 
Initiate new segment


 99%|█████████▉| 7849/7944 [01:07<00:00, 119.70it/s]

Change detected at: 8024 
Initiate new segment
Change detected at: 8031 
Initiate new segment


100%|█████████▉| 7909/7944 [01:07<00:00, 115.93it/s]

Change detected at: 8087 
Initiate new segment


100%|██████████| 7944/7944 [01:08<00:00, 116.60it/s]


F1 score: 0.5161290322580646
Covering: 0.7793320648897838
0.5299145299145299
cuda


  8%|▊         | 652/7944 [00:05<01:00, 120.07it/s]

Change detected at: 828 
Initiate new segment


 16%|█▌        | 1290/7944 [00:10<00:52, 126.66it/s]

Change detected at: 1474 
Initiate new segment


 26%|██▋       | 2091/7944 [00:18<03:51, 25.33it/s] 

Change detected at: 2270 
Initiate new segment


 30%|███       | 2412/7944 [00:21<00:47, 117.34it/s]

Change detected at: 2594 
Initiate new segment


 32%|███▏      | 2514/7944 [00:21<00:44, 123.09it/s]

Change detected at: 2697 
Initiate new segment


 35%|███▍      | 2743/7944 [00:23<00:41, 126.83it/s]

Change detected at: 2922 
Initiate new segment


 42%|████▏     | 3373/7944 [00:28<00:35, 129.67it/s]

Change detected at: 3554 
Initiate new segment


 45%|████▍     | 3572/7944 [00:29<00:32, 132.78it/s]

Change detected at: 3755 
Initiate new segment
Change detected at: 3760 
Initiate new segment


 49%|████▉     | 3909/7944 [00:32<00:30, 133.62it/s]

Change detected at: 4087 
Initiate new segment


 61%|██████    | 4824/7944 [00:39<00:25, 120.06it/s]

Change detected at: 5003 
Initiate new segment
Change detected at: 5017 
Initiate new segment


 63%|██████▎   | 5014/7944 [00:41<00:23, 123.39it/s]

Change detected at: 5193 
Initiate new segment


 66%|██████▋   | 5271/7944 [00:43<00:26, 102.24it/s]

Change detected at: 5455 
Initiate new segment


 68%|██████▊   | 5422/7944 [00:45<00:24, 104.56it/s]

Change detected at: 5606 
Initiate new segment


 80%|███████▉  | 6318/7944 [00:55<00:15, 106.93it/s]

Change detected at: 6499 
Initiate new segment
Change detected at: 6507 
Initiate new segment


 81%|████████  | 6429/7944 [00:56<00:14, 105.99it/s]

Change detected at: 6607 
Initiate new segment


 87%|████████▋ | 6906/7944 [01:00<00:08, 118.31it/s]

Change detected at: 7086 
Initiate new segment


 90%|████████▉ | 7111/7944 [01:02<00:07, 114.26it/s]

Change detected at: 7298 
Initiate new segment


 99%|█████████▉| 7869/7944 [01:08<00:00, 114.89it/s]

Change detected at: 8054 
Initiate new segment


100%|█████████▉| 7905/7944 [01:09<00:00, 108.15it/s]

Change detected at: 8087 
Initiate new segment


100%|██████████| 7944/7944 [01:09<00:00, 114.24it/s]


F1 score: 0.4444444444444444
Covering: 0.7584580304660872
0.4816053511705686
cuda


  6%|▌         | 450/7744 [00:03<01:03, 114.70it/s]

Change detected at: 828 
Initiate new segment


 15%|█▍        | 1161/7744 [00:10<00:56, 116.84it/s]

Change detected at: 1538 
Initiate new segment


 24%|██▍       | 1893/7744 [00:18<00:56, 104.23it/s]

Change detected at: 2270 
Initiate new segment


 34%|███▎      | 2603/7744 [00:24<00:45, 113.37it/s]

Change detected at: 2990 
Initiate new segment


 42%|████▏     | 3253/7744 [00:30<00:38, 116.94it/s]

Change detected at: 3630 
Initiate new segment


 44%|████▎     | 3387/7744 [00:31<00:39, 110.85it/s]

Change detected at: 3770 
Initiate new segment


 63%|██████▎   | 4845/7744 [00:44<00:26, 111.20it/s]

Change detected at: 5234 
Initiate new segment


 66%|██████▌   | 5090/7744 [00:46<00:22, 118.57it/s]

Change detected at: 5470 
Initiate new segment


 81%|████████  | 6278/7744 [00:59<00:11, 122.95it/s]

Change detected at: 6655 
Initiate new segment


 91%|█████████ | 7038/7744 [01:05<00:05, 130.50it/s]

Change detected at: 7420 
Initiate new segment


 99%|█████████▉| 7701/7744 [01:10<00:00, 109.67it/s]

Change detected at: 8087 
Initiate new segment


100%|██████████| 7744/7744 [01:11<00:00, 108.91it/s]


F1 score: 0.48000000000000004
Covering: 0.7413632016488939
0.4807692307692307
cuda


  6%|▌         | 439/7744 [00:04<01:09, 104.90it/s]

Change detected at: 828 
Initiate new segment


 15%|█▌        | 1168/7744 [00:10<01:01, 107.73it/s]

Change detected at: 1550 
Initiate new segment


 24%|██▍       | 1885/7744 [00:20<01:14, 78.83it/s] 

Change detected at: 2270 
Initiate new segment


 34%|███▎      | 2601/7744 [00:28<00:56, 91.63it/s] 

Change detected at: 2990 
Initiate new segment


 42%|████▏     | 3286/7744 [00:35<00:48, 92.66it/s] 

Change detected at: 3671 
Initiate new segment


 63%|██████▎   | 4848/7744 [00:54<00:29, 97.54it/s] 

Change detected at: 5230 
Initiate new segment


 81%|████████  | 6273/7744 [01:09<00:13, 105.25it/s]

Change detected at: 6655 
Initiate new segment


 90%|█████████ | 7004/7744 [01:17<00:07, 103.70it/s]

Change detected at: 7393 
Initiate new segment


 99%|█████████▉| 7703/7744 [01:26<00:00, 98.74it/s] 

Change detected at: 8088 
Initiate new segment


100%|██████████| 7744/7744 [01:26<00:00, 89.22it/s]


F1 score: 0.4347826086956522
Covering: 0.7275248609379484
0.44230769230769235
cuda


  5%|▍         | 344/7644 [00:03<01:12, 100.28it/s]

Change detected at: 828 
Initiate new segment


 14%|█▎        | 1048/7644 [00:11<01:21, 80.80it/s] 

Change detected at: 1535 
Initiate new segment


 23%|██▎       | 1783/7644 [00:19<01:09, 84.18it/s] 

Change detected at: 2271 
Initiate new segment


 33%|███▎      | 2493/7644 [00:26<00:55, 92.91it/s] 

Change detected at: 2978 
Initiate new segment


 42%|████▏     | 3216/7644 [00:36<00:50, 88.13it/s] 

Change detected at: 3704 
Initiate new segment


 62%|██████▏   | 4749/7644 [00:52<00:29, 99.65it/s] 

Change detected at: 5234 
Initiate new segment


 81%|████████  | 6171/7644 [01:09<00:15, 98.18it/s] 

Change detected at: 6655 
Initiate new segment


 91%|█████████▏| 6980/7644 [01:17<00:06, 100.19it/s]

Change detected at: 7468 
Initiate new segment


100%|█████████▉| 7611/7644 [01:23<00:00, 90.74it/s] 

Change detected at: 8097 
Initiate new segment


100%|██████████| 7644/7644 [01:24<00:00, 90.69it/s]


F1 score: 0.4347826086956522
Covering: 0.726468170063331
0.44230769230769235
cuda


  4%|▍         | 342/7644 [00:03<01:12, 100.87it/s]

Change detected at: 829 
Initiate new segment


 14%|█▍        | 1080/7644 [00:10<01:06, 98.11it/s]

Change detected at: 1563 
Initiate new segment


 23%|██▎       | 1791/7644 [00:19<00:54, 107.98it/s]

Change detected at: 2272 
Initiate new segment


 33%|███▎      | 2533/7644 [00:27<00:47, 107.07it/s]

Change detected at: 3015 
Initiate new segment


 43%|████▎     | 3275/7644 [00:34<00:44, 97.81it/s] 

Change detected at: 3760 
Initiate new segment


 62%|██████▏   | 4748/7644 [00:51<00:27, 103.55it/s]

Change detected at: 5236 
Initiate new segment


 81%|████████  | 6175/7644 [01:07<00:15, 94.22it/s] 

Change detected at: 6655 
Initiate new segment


 92%|█████████▏| 7008/7644 [01:15<00:05, 108.75it/s]

Change detected at: 7493 
Initiate new segment


100%|█████████▉| 7620/7644 [01:23<00:00, 95.65it/s] 

Change detected at: 8108 
Initiate new segment


100%|██████████| 7644/7644 [01:24<00:00, 90.89it/s]

F1 score: 0.5217391304347826
Covering: 0.7111024746310426
0.5307692307692308


In [38]:
FPs

[30, 28, 9, 14, 6, 5, 5, 4]

In [36]:
f1,aUC,cov,fp, dd = detector.evaluate(GroundTruth,30)
fp

F1 score: 0.5833333333333334
Covering: 0.8032426077984104
0.5874125874125874


4

In [14]:
SWD_AUCS = [detector.evaluate(GroundTruth,tau)[2] for tau in [10,30,50,75,100]]
SWD_AUCS

F1 score: 0.41666666666666663
Covering: 0.8027825975884822
0.4195804195804196
F1 score: 0.5833333333333334
Covering: 0.8027825975884822
0.5874125874125874
F1 score: 0.6666666666666667
Covering: 0.8027825975884822
0.6713286713286712
F1 score: 0.6666666666666667
Covering: 0.8027825975884822
0.6713286713286712
F1 score: 0.8333333333333333
Covering: 0.8027825975884822
0.8391608391608393


[0.4195804195804196,
 0.5874125874125874,
 0.6713286713286712,
 0.6713286713286712,
 0.8391608391608393]

### Clasp

In [11]:
from claspy.segmentation import BinaryClaSPSegmentation

clasp = BinaryClaSPSegmentation(window_size=50).fit(Occupancy.to_numpy())
print('F1 & AUC & FP',f_measure({'0':GroundTruth},clasp.change_points,30))
covering({'0':GroundTruth},clasp.change_points,Occupancy.shape[0])
detection_delay(GroundTruth,clasp.change_points)

F1 & AUC & FP (0.2666666666666667, 0.5769230769230769, 0)


({828: 3}, 3.0)

In [12]:
CLASP_AUCS = [f_measure({'0':GroundTruth},clasp.change_points,tau)[2] for tau in [10,30,50,75,100]]
CLASP_AUCS

[0, 0, 0, 0, 0]

### Results for R methods

In [6]:
import json

#R=30,sig.lvl=0.05,min.size = 400
with open("./R/res/OccupancyECP.json") as f:
    res = json.load(f)[1:-1]

ECP_CPS =res
print("f1 & AUC & FP",f_measure({'0':GroundTruth},ECP_CPS,30))
print("Covering",covering({'0':GroundTruth},ECP_CPS,Occupancy.shape[0]))
print('Delay', detection_delay(GroundTruth,ECP_CPS)[1])

ECP_AUCS = [f_measure({'0':GroundTruth},ECP_CPS,tau)[1] for tau in [10,30,50,75,100]]
#BOCPD with datanormalization
with open("./R/res/OccupancyBOCPD.json") as f:
    res = json.load(f)[0][1:]

BOCPD_CPS =  res
print("F1 & AUC & FP",f_measure({'0':GroundTruth},BOCPD_CPS,30))
print("Covering",covering({'0':GroundTruth},BOCPD_CPS,Occupancy.shape[0]))
print('Delay',detection_delay(GroundTruth,BOCPD_CPS)[1])

BOCPD_AUCS = [f_measure({'0':GroundTruth},BOCPD_CPS,tau)[1] for tau in [10,30,50,75,100]]


print(ECP_AUCS)
print(BOCPD_AUCS)

f1 & AUC & FP (0.33333333333333337, 0.3393665158371041, 12)
Covering 0.6401316828686778
Delay 167.4375
F1 & AUC & FP (0.5454545454545455, 0.5711538461538461, 11)
Covering 0.7367589017005471
Delay 105.6842105263158
[0.2714932126696833, 0.3393665158371041, 0.40723981900452494, 0.6108597285067873, 0.6787330316742082]
[0.5711538461538461, 0.5711538461538461, 0.5711538461538461, 0.6346153846153847, 0.698076923076923]


### RULSIF

In [16]:
import numpy as np
import pandas as pd
from densratio import densratio
#from sklearn.preprocessing import StandardScaler
from matplotlib import pyplot as plt
from tqdm import tqdm

# --- 1. Define the Sliding Window Function ---

def calculate_rulsif_scores(
    time_series_data, # (N_timesteps, N_features) numpy array
    window_size,
    step_size,
    alpha=0.1
):
    """
    Applies RuLSIF using a two-window sliding approach for change detection.

    Args:
        time_series_data (np.ndarray): The multivariate time series data.
        window_size (int): The size of the reference (x1) and test (x2) windows.
        step_size (int): The number of steps to slide the windows forward.
        alpha (float): The relative density parameter for RuLSIF (0.1 is standard).

    Returns:
        tuple: (list of window center indices, list of RPD scores)
    """
    N, D = time_series_data.shape
    rpd_scores = []
    window_indices = []
    
    # The comparison starts after the initial reference window is established
    start_index = window_size
    
    # Loop to slide the two windows (Reference and Test)
    # The total window span is 2 * window_size. The first score is registered 
    # at the end of the Test Window (which is the potential change-point)
    for i in tqdm(range(start_index, N - window_size, step_size)):
        
        # 1. Define the Reference Window (x1) - Past data
        # From i - window_size to i
        x1 = time_series_data[i - window_size : i, :]
        
        # 2. Define the Test Window (x2) - Current data
        # From i to i + window_size
        x2 = time_series_data[i : i + window_size, :]
        
        # Check if windows are of the correct size (safety check)
        if x1.shape[0] != window_size or x2.shape[0] != window_size:
            continue
            
        try:
            # 3. Apply RuLSIF
            # x1 is the numerator (p(x)), x2 is the denominator (q(x))
            densratio_obj = densratio(
                x1, x2, 
                alpha=alpha,
                sigma_range="auto", 
                lambda_range="auto", verbose=False
            )
            
            # The alpha-relative PE divergence (RPD) is a measure of dissimilarity
            # between the distributions p(x) and q(x).
            score = densratio_obj.alpha_PE
            rpd_scores.append(score)
            
            # The score is logged at the end of the Test Window
            window_indices.append(i + window_size) 
            
        except Exception as e:
            # Handle cases where optimization might fail (rare with good data)
            print(f"RuLSIF error at index {i}: {e}")
            rpd_scores.append(np.nan)
            window_indices.append(i + window_size)
        #print(densratio_obj.alpha_PE)  
    return np.array(window_indices), np.array(rpd_scores)

signal = (Occupancy-Occupancy.mean())/Occupancy.std()
signal = signal.values.astype(np.float32)

detection_RULIF = calculate_rulsif_scores(signal[:,0][:,np.newaxis],20,step_size=2)

  1%|          | 33/4052 [00:01<02:52, 23.34it/s]/home/sven/miniconda3/lib/python3.12/site-packages/densratio/RuLSIF.py:260: RuntimeWarning: divide by zero encountered in log
  divergence = np.log(g_x).sum(axis=0) / n
100%|██████████| 4052/4052 [03:20<00:00, 20.24it/s]


## FP

### Results

| METHOD  | DD  | FP  |
|-------------|----------|-----------|
| SWD (ours)       | 52 | 4|
| ECP         | 167 | 12  |
| KCPA         |77 | 11|
| CLasP         | 3 | 0 |
| BOCPD         | 105  | 11 |